File: src/main/java/szewek/flux/FBlocks.java
package szewek.flux;

import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraftforge.registries.IForgeRegistry;
import szewek.flux.block.*;
import szewek.flux.util.Metal;

import java.util.EnumMap;
import java.util.Map;

import static szewek.flux.FluxMod.MODID;

public final class FBlocks {
	static final Map<Metal, FluxOreBlock> ORES = makeOres();
	static final Map<Metal, MetalBlock> METAL_BLOCKS = makeBlocks();

	public static final FluxGenBlock FLUXGEN = new FluxGenBlock(Block.Properties.create(Material.IRON).hardnessAndResistance(1f));
	public static final EnergyCableBlock ENERGY_CABLE = new EnergyCableBlock(Block.Properties.create(Material.IRON).hardnessAndResistance(0.3f));
	public static final MachineBlock
			GRINDING_MILL = new MachineBlock(() -> FTiles.GRINDING_MILL),
			ALLOY_CASTER = new MachineBlock(() -> FTiles.ALLOY_CASTER),
			WASHER = new MachineBlock(() -> FTiles.WASHER),
			COMPACTOR = new MachineBlock(() -> FTiles.COMPACTOR);

	static void register(final IForgeRegistry<Block> reg) {
		ORES.values().forEach(reg::register);
		METAL_BLOCKS.values().forEach(reg::register);
		reg.registerAll(
				FLUXGEN.setRegistryName(MODID, "fluxgen"),
				ENERGY_CABLE.setRegistryName(MODID, "energy_cable"),
				GRINDING_MILL.setRegistryName(MODID, "grinding_mill"),
				ALLOY_CASTER.setRegistryName(MODID, "alloy_caster"),
				WASHER.setRegistryName(MODID, "washer"),
				COMPACTOR.setRegistryName(MODID, "compactor")
		);
	}

	private static Map<Metal, FluxOreBlock> makeOres() {
		Map<Metal, FluxOreBlock> m = new EnumMap<>(Metal.class);
		for (Metal metal : Metal.values()) {
			if (Metal.nonVanilla(metal)) {
				FluxOreBlock b = new FluxOreBlock(metal);
				b.setRegistryName(MODID, metal.name + "_ore");
				m.put(metal, b);
			}
		}
		return m;
	}

	private static Map<Metal, MetalBlock> makeBlocks() {
		Map<Metal, MetalBlock> m = new EnumMap<>(Metal.class);
		for (Metal metal : Metal.values()) {
			if (Metal.nonVanilla(metal)) {
				MetalBlock b = new MetalBlock(Block.Properties.create(Material.IRON), metal);
				b.setRegistryName(MODID, metal.name + "_block");
				m.put(metal, b);
			}
		}
		return m;
	}
}

File: src/main/java/szewek/flux/FBlocks.kt
package szewek.flux

import net.minecraft.block.Block
import net.minecraft.block.SoundType
import net.minecraft.block.material.Material
import net.minecraftforge.registries.IForgeRegistry
import szewek.flux.block.*
import szewek.flux.util.Metal
import szewek.ktutils.blockProps
import java.util.*

object FBlocks {
    @JvmField val ORES = makeOres()
    @JvmField val METAL_BLOCKS = makeBlocks()
    @JvmField val FLUXGEN = FluxGenBlock(blockProps(Material.IRON) {
        hardnessAndResistance(1f)
        sound(SoundType.METAL)
    })
    @JvmField val ENERGY_CABLE = EnergyCableBlock(blockProps(Material.IRON) {
        hardnessAndResistance(0.3f)
    })
    @JvmField val DIGGER = DiggerBlock(blockProps(Material.IRON) {
        hardnessAndResistance(1f)
        sound(SoundType.METAL)
    })
    @JvmField val GRINDING_MILL = MachineBlock()
    @JvmField val ALLOY_CASTER = MachineBlock()
    @JvmField val WASHER = MachineBlock()
    @JvmField val COMPACTOR = MachineBlock()
    @JvmStatic
    fun register(reg: IForgeRegistry<Block?>) {
        ORES.values.forEach { value: FluxOreBlock? -> reg.register(value) }
        METAL_BLOCKS.values.forEach { value: MetalBlock? -> reg.register(value) }
        reg.registerAll(
                FLUXGEN.setRegistryName(FluxMod.MODID, "fluxgen"),
                ENERGY_CABLE.setRegistryName(FluxMod.MODID, "energy_cable"),
                DIGGER.setRegistryName(FluxMod.MODID, "digger"),
                GRINDING_MILL.setRegistryName(FluxMod.MODID, "grinding_mill"),
                ALLOY_CASTER.setRegistryName(FluxMod.MODID, "alloy_caster"),
                WASHER.setRegistryName(FluxMod.MODID, "washer"),
                COMPACTOR.setRegistryName(FluxMod.MODID, "compactor")
        )
    }

    private fun makeOres(): Map<Metal, FluxOreBlock> {
        val m: MutableMap<Metal, FluxOreBlock> = EnumMap<Metal, FluxOreBlock>(Metal::class.java)
        for (metal in Metal.values()) {
            if (metal.nonVanilla()) {
                val b = FluxOreBlock(metal)
                b.setRegistryName(FluxMod.MODID, metal.metalName + "_ore")
                m[metal] = b
            }
        }
        return m
    }

    private fun makeBlocks(): Map<Metal, MetalBlock> {
        val m: MutableMap<Metal, MetalBlock> = EnumMap<Metal, MetalBlock>(Metal::class.java)
        for (metal in Metal.values()) {
            if (metal.nonVanilla()) {
                val b = MetalBlock(metal)
                b.setRegistryName(FluxMod.MODID, metal.metalName + "_block")
                m[metal] = b
            }
        }
        return m
    }
}